### ISSUE TITLE
ENH: refactor our loop implementations into a stand-alone library

### ISSUE THREAD

**Issue by mattip** [numpy/numpy#17698]
We are slowly getting to a best-of-breed set of CPU inner loop functions using Unversal Intrinsics. It would be helpful to other projects if we could refactor these into a stand-alone C/C++ library, so they could be re-used in the growing number of NumPy-like libraries.

**Comment by Qiyu8** [comment]
What a coincidence, I started a cross platform portable [project](https://github.com/mathlibrary/usimd) months ago, I don't know if this is what you want, The basic idea is extracting the current numpy universal intrinsics framework into a standalone project that can benefit other numpy-like scientific libraries. Currently this project is a prototype, you can have fun by running [this script](https://github.com/mathlibrary/usimd/blob/master/test/runAllTests.sh).

**Comment by seiko2plus** [comment]
@mattip,

> It would be helpful to other projects if we could refactor these into a stand-alone C/C++ library

I disagree, other projects have to create their own framework or build their own on top of ours. We had discussed this before in #13516. General-purpose SIMD projects aren't good for the long term. usually, these kinda libraries going to swim in an ocean of SIMD kernels trying to satisfy everybody, and in the end, big projects soon or later will get rid of it. I can name to you a dozen of SIMD libraries that suffered the same fate.

**Comment by mattip** [comment]
There is a defined set of element-wise functions in the [array API standard](https://data-apis.github.io/array-api/latest/API_specification/elementwise_functions.html) and @Qiyu8 has started [mathlibrary/usimd](https://github.com/mathlibrary/usimd) extracted from the USIMD work. I agree general purpose SIMD projects need fixed boundaries, the Array API work gives us that framework.

**Comment by seiko2plus** [comment]
Too hard to put a limit on people's optimization needs, it always starts like that but the workflow will stretch it. for example:
 - functions with different ULPS
 - with/without domain/fp raising errors
 - handling both non-contiguous/contiguous memory access

Multiply all the above points by the number of supported data types, if all of that okay for you then prepare yourself for the main
issue "how you going to manage the relationship between static(--cpu-baseline) and dynamic(--cpu-dispatch) dispatching?" most of the SIMD libraries escape from this issue by exporting the SIMD kernels as inline functions and let the users manage their own build, good solution but only if you are using C++ since templates will reduce the amount of code.
